doc/specs/stdlib_sorting.md

```diff
@@ -0,0 +1,644 @@
+argument or allocated internally on the stack.
+
+#### The `SORT_INDEX` subroutine
+
+The `SORT` and `ORD_SORT` subroutines can sort rank 1 isolated
+arrays of intrinsic types, but do nothing for the coordinated sorting
+of related data, e.g., multiple related rank 1 arrays, higher rank
+arrays, or arrays of derived types. For such related data, what is
+useful is an array of indices that maps a rank 1 array to its sorted
+form. For such a sort, a stable sort is useful, therefore the module
+provides a subroutine, `SORT_INDEX`, that generates such an array of
+indices based on the `ORD_SORT` algorithm, in addition to sorting
+the input array.
+
+The logic of `SORT_INDEX` parallels that of `ORD_SORT`, with
+additional housekeeping to keep the array of indices consistent with
+the sorted positions of the input array. Because of this additional
+housekeeping it has slower runtime performance than `ORD_SORT`.
+`SORT_INDEX` requires the use of two "scratch" arrays, that may be
+provided as optional `work` and `iwork` arguments or allocated
+internally on the stack.
+
+#### The `SORT` subroutine
+
+`SORT` uses the `introsort` sorting algorithm of David Musser.
+`introsort` is a hybrid unstable comparison algorithm combining
+`quicksort`, `insertion sort`, and `heap sort`. While this algorithm's
+runtime performance is always O(N Ln(N)), it is relatively fast on
+randomly ordered data, but does not show the improvement in
+performance on partly sorted data found for `ORD_SORT`.
+
+As with `introsort`, `SORT` is an unstable hybrid algorithm.
+First it examines the array and estimates the depth of recursion a
+quick sort would require for ideal (random) data, `D =
+Ceiling(Ln(N)/Ln(2))`. It then defines a limit to the number of
+`quicksort` recursions to be allowed in processing,
+`D_limit = factor * D`, where factor is currently 2, and
+calls `introsort` proper. `introsort` proper then:
+
+1. Examines the number of elements remaining to be sorted, and, if
+   they are less than 16, sorts them using insertion sort and returns;
+2. If they are not less than 16, checks whether the current depth of
+   recursion exceeds `D_limit` and, if it does, processes the remaining
+   elements with heap sort and returns;
+3. If the current depth of recursion does not exceed `D_limit`, then
+   in effect does a `quicksort` step:
+
+    * Partitions the remaining array using a median of three,
+    * Calls `introsort` proper on the leftmost partition,
+    * Calls `introsort` proper on the rightmost partition, and then
+      returns.
+
+The resulting algorithm is of order O(N Ln(N)) run time performance
+for all inputs. Because it relies on `quicksort`, the coefficient of
+the O(N Ln(N)) behavior is typically small compared to other sorting
+algorithms on random data. On partially sorted data it can show either
+slower `heap sort` performance, or enhanced performance by up to a
+factor of six. Still, even when it shows enhanced performance, its
```

Comment:
"factor of six": based on your benchmarks?

Reply:
Yes. For example, with the Intel compiler the three truly"Random" benchmarks (Random dense, Random Order, and Random Sparse) take about 0.071 seconds while the "Increasing" benchmark takes about 0.012 seconds almost a factor of six faster. However for the gfortran tests, the improvement is only about a factor of three.